{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bea51f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:27883/\n",
      "\n",
      " * Serving Flask app \"__main__\" (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " * Running on http://127.0.0.1:27883/ (Press CTRL+C to quit)\n",
      "127.0.0.1 - - [23/Feb/2024 21:42:55] \"GET / HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:42:56] \"GET /_dash-dependencies HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:42:58] \"GET /_dash-layout HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:42:59] \"GET /_dash-component-suites/dash/dash_table/async-highlight.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:42:59] \"GET /_dash-component-suites/dash/dash_table/async-table.js HTTP/1.1\" 304 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 166, in update_dashboard\n",
      "    df = pd.DataFrame.from_records(animals.read({}))\n",
      "NameError: name 'animals' is not defined\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:42:59] \"POST /_dash-update-component HTTP/1.1\" 500 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 212, in update_map\n",
      "    row_ids[0] = row_ids[len(row_ids) - 1]\n",
      "TypeError: object of type 'NoneType' has no len()\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:42:59] \"POST /_dash-update-component HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:42:59] \"POST /_dash-update-component HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 200, in update_graphs\n",
      "    figure = px.histogram(dff, x='breed')\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/plotly/express/_chart_types.py\", line 472, in histogram\n",
      "    return make_figure(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/plotly/express/_core.py\", line 1945, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/plotly/express/_core.py\", line 1405, in build_dataframe\n",
      "    df_output, wide_id_vars = process_args_into_dataframe(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/plotly/express/_core.py\", line 1207, in process_args_into_dataframe\n",
      "    raise ValueError(err_msg)\n",
      "ValueError: Value of 'x' is not the name of a column in 'data_frame'. Expected one of [] but received: breed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:42:59] \"POST /_dash-update-component HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:43:00] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:43:00] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:43:00] \"GET /_dash-component-suites/dash/dcc/async-graph.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:43:00] \"GET /_dash-component-suites/dash/dcc/async-plotlyjs.js HTTP/1.1\" 304 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:43:57] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:44:00] \"POST /_dash-update-component HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 212, in update_map\n",
      "    row_ids[0] = row_ids[len(row_ids) - 1]\n",
      "IndexError: list index out of range\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:44:01] \"POST /_dash-update-component HTTP/1.1\" 500 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 155, in update_dashboard\n",
      "    df = pd.DataFrame.from_records(animals.read({\n",
      "NameError: name 'animals' is not defined\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:45:02] \"POST /_dash-update-component HTTP/1.1\" 500 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 144, in update_dashboard\n",
      "    df = pd.DataFrame.from_records(animals.read({\n",
      "NameError: name 'animals' is not defined\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:45:07] \"POST /_dash-update-component HTTP/1.1\" 500 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 135, in update_dashboard\n",
      "    df = pd.DataFrame.from_records(animals.read({\n",
      "NameError: name 'animals' is not defined\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:45:09] \"POST /_dash-update-component HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:45:56] \"POST /_dash-update-component HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:45:59] \"POST /_dash-update-component HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 144, in update_dashboard\n",
      "    df = pd.DataFrame.from_records(animals.read({\n",
      "NameError: name 'animals' is not defined\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:46:02] \"POST /_dash-update-component HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [23/Feb/2024 21:46:06] \"POST /_dash-update-component HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 212, in update_map\n",
      "    row_ids[0] = row_ids[len(row_ids) - 1]\n",
      "IndexError: list index out of range\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:46:07] \"POST /_dash-update-component HTTP/1.1\" 500 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/dash.py\", line 1273, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/anaconda/lib/python3.9/site-packages/dash/_callback.py\", line 440, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"/tmp/ipykernel_286333/830668042.py\", line 155, in update_dashboard\n",
      "    df = pd.DataFrame.from_records(animals.read({\n",
      "NameError: name 'animals' is not defined\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [23/Feb/2024 21:46:08] \"POST /_dash-update-component HTTP/1.1\" 500 -\n"
     ]
    }
   ],
   "source": [
    "# Setup the Jupyter version of Dash\n",
    "from jupyter_dash import JupyterDash\n",
    "\n",
    "# Configure the necessary Python module imports\n",
    "import dash\n",
    "import dash_leaflet as dl\n",
    "from dash import dcc\n",
    "from dash import html\n",
    "import plotly.express as px\n",
    "from dash import dash_table\n",
    "from dash.dependencies import Input, Output\n",
    "import base64\n",
    "\n",
    "# Configure the plotting routines\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "#### FIX ME #####\n",
    "# change animal_shelter and AnimalShelter to match your CRUD Python module file name and class name\n",
    "from AnimalShelter import AnimalShelter\n",
    "\n",
    "\n",
    "\n",
    "###########################\n",
    "# Data Manipulation / Model\n",
    "###########################\n",
    "# FIX ME update with your username and password and CRUD Python module name. NOTE: You will\n",
    "# likely need more variables for your constructor to handle the hostname and port of the MongoDB\n",
    "# server, and the database and collection names\n",
    "\n",
    "def __init__(self, client, database, collection):\n",
    "    self.client = MongoClient('mongodb://%s:%s@nv-desktop-services.apporto.com:30275' % (\"aacuser\", \"SNHU1234\"))\n",
    "    self.database = self.client['AAC']\n",
    "    self.collection = self.database['animals']\n",
    "\n",
    "shelter = AnimalShelter(\"aacuser\",\"SNHU1234\",\"AAC\")\n",
    "\n",
    "# class read method must support return of list object and accept projection json input\n",
    "# sending the read method an empty document requests all documents be returned\n",
    "df = pd.DataFrame.from_records(shelter.read({}))\n",
    "\n",
    "# MongoDB v5+ is going to return the '_id' column and that is going to have an \n",
    "# invlaid object type of 'ObjectID' - which will cause the data_table to crash - so we remove\n",
    "# it in the dataframe here. The df.drop command allows us to drop the column. If we do not set\n",
    "# inplace=True - it will reeturn a new dataframe that does not contain the dropped column(s)\n",
    "df.drop(columns=['_id'],inplace=True)\n",
    "\n",
    "## Debug\n",
    "# print(len(df.to_dict(orient='records')))\n",
    "# print(df.columns)\n",
    "\n",
    "\n",
    "#########################\n",
    "# Dashboard Layout / View\n",
    "#########################\n",
    "external_stylesheets = ['https://codepen.io/chriddyp/pen/bWLwgP.css']\n",
    "app = dash.Dash(__name__, external_stylesheets=external_stylesheets)\n",
    "\n",
    "\n",
    "image_filename = 'Grazioso Salvare Logo.png' # replace with your own image\n",
    "encoded_image = base64.b64encode(open(image_filename, 'rb').read())\n",
    "                               \n",
    "app.layout = html.Div([\n",
    "    html.Div(id='hidden-div', style={'display':'none'}),\n",
    "    html.Center(html.B(html.H1('SNHU CS-340 Dashboard'))),\n",
    "    html.Center(html.B(html.H1('Taylor Devine Project 2 Dashboard'))),\n",
    "    html.Hr(),\n",
    "    html.Img(id='customer-image',src='data:image/png;base64,{}'.format(encoded_image.decode()),alt='customer image'),\n",
    "    html.Div(      \n",
    "        dcc.RadioItems(\n",
    "            id='filter-type',\n",
    "            options=[\n",
    "                {'label': 'Water Rescue', 'value': 'water'},\n",
    "                {'label': 'Mountain/Wilderness Rescue', 'value': 'mount'},\n",
    "                {'label': 'Disaster Rescue and Individual Tracking', 'value': 'disaster'},\n",
    "                {'label': 'Reset', 'value': 'reset'}\n",
    "            ],\n",
    "            value='reset'\n",
    "        )\n",
    "        \n",
    "    ),\n",
    "    html.Hr(),\n",
    "    dash_table.DataTable(\n",
    "        id='datatable-id',\n",
    "        columns=[\n",
    "            {\"name\": i, \"id\": i, \"deletable\": False, \"selectable\": True} for i in df.columns\n",
    "        ],\n",
    "        data=df.to_dict('records'),\n",
    "        # Allows 10 rows per page, and starts on page 0. Columns and Rows may be selected but not deleted/edited\n",
    "        editable=False,\n",
    "        sort_action=\"native\",\n",
    "        sort_mode=\"multi\",\n",
    "        column_selectable=True,\n",
    "        row_selectable=True,\n",
    "        row_deletable=False,\n",
    "        selected_columns=[],\n",
    "        selected_rows=[0],\n",
    "        page_action=\"native\",\n",
    "        page_current= 0,\n",
    "        page_size= 10,\n",
    "    ),\n",
    "    html.Br(),\n",
    "    html.Hr(),\n",
    "#This sets up the dashboard so that your chart and your geolocation chart are side-by-side\n",
    "    html.Div(className='row',\n",
    "         style={'display' : 'flex'},\n",
    "             children=[\n",
    "        html.Div(\n",
    "            id='graph-id',\n",
    "            className='col s12 m6',\n",
    "\n",
    "            ),\n",
    "        html.Div(\n",
    "            id='map-id',\n",
    "            className='col s12 m6',\n",
    "            )\n",
    "        ])\n",
    "])\n",
    "\n",
    "#############################################\n",
    "# Interaction Between Components / Controller\n",
    "#############################################\n",
    "\n",
    "\n",
    "\n",
    "# Radio button filters\n",
    "@app.callback([Output('datatable-id','data'),\n",
    "               Output('datatable-id','columns')],\n",
    "              [Input('filter-type', 'value')])\n",
    "def update_dashboard(filter_type):\n",
    "        # Water Rescue filter\n",
    "        if filter_type == 'water':\n",
    "            df = pd.DataFrame.from_records(animals.read({\n",
    "                \"animal_type\": \"Dog\",\n",
    "                \"breed\": {\"$in\": [\"Labrador Retriever Mix\",\"Chesapeake Bay Retriever\", \"Newfoundland\"\n",
    "                                 ]},\n",
    "                \"sex_upon_outcome\": \"Intact Female\",\n",
    "                \"age_upon_outcome_in_weeks\": {\"$gte\":26.0, \"$lte\":156.0}\n",
    "            }))\n",
    "        # Mountain or Wilderness Rescue filter\n",
    "        elif filter_type == 'mount':\n",
    "            df = pd.DataFrame.from_records(animals.read({\n",
    "                \"animal_type\": \"Dog\",\n",
    "                \"breed\": {\"$in\": [\"German Shepard\",\"Alaskan Malamute\",\"Old English Sheepdog\", \n",
    "                                  \"Siberian Husky\", \"Rottweiler\"\n",
    "                                 ]},\n",
    "                \"sex_upon_outcome\": \"Intact Male\",\n",
    "                \"age_upon_outcome_in_weeks\": {\"$gte\":26.0, \"$lte\":156.0}\n",
    "            }))\n",
    "            \n",
    "        # Disaster Rescue or Individual Tracking filter\n",
    "        elif filter_type == 'disaster':\n",
    "            df = pd.DataFrame.from_records(animals.read({\n",
    "                \"animal_type\": \"Dog\",\n",
    "                \"breed\": {\"$in\": [\"Doberman Pinscher\",\"German Shepard\",\"Golden Retriever\", \n",
    "                                  \"Bloodhound\",\"Rottweiler\"\n",
    "                                 ]},\n",
    "                \"sex_upon_outcome\": \"Intact Male\",\n",
    "                \"age_upon_outcome_in_weeks\": {\"$gte\":20.0, \"$lte\":300.0}\n",
    "            }))\n",
    "            \n",
    "        # Reset to no filter\n",
    "        else:\n",
    "            df = pd.DataFrame.from_records(animals.read({}))\n",
    "        \n",
    "        columns=[{\"name\": i, \"id\": i, \"deletable\": False, \"selectable\": True} for i in df.columns]\n",
    "        data=df.to_dict('records')\n",
    "        \n",
    "        \n",
    "        return (data,columns)\n",
    "\n",
    "# Change background color of selected columns\n",
    "@app.callback(\n",
    "    Output('datatable-id', 'style_data_conditional'),\n",
    "    [Input('datatable-id', 'selected_columns')]\n",
    ")\n",
    "def update_styles(selected_columns):\n",
    "    return [{\n",
    "        'if': { 'column_id': i },\n",
    "        'background_color': '#D2F3FF'\n",
    "    } for i in selected_columns]\n",
    "\n",
    "# Chart\n",
    "@app.callback(\n",
    "    Output('graph-id', \"children\"),\n",
    "    [Input('datatable-id', \"derived_viewport_data\"),\n",
    "     #Input('filter-type', 'value')\n",
    "    ])\n",
    "def update_graphs(viewData):\n",
    "    ###FIX ME ####\n",
    "    # add code for chart of your choice (e.g. pie chart) #\n",
    "    dff = pd.DataFrame.from_dict(viewData)\n",
    "    \n",
    "    return [\n",
    "        dcc.Graph(            \n",
    "#             figure = px.pie(dff, values='breed',  \n",
    "#                             title='Available Dogs by Breed')\n",
    "            figure = px.histogram(dff, x='breed')\n",
    "        )    \n",
    "    ]\n",
    "\n",
    "# Map\n",
    "@app.callback(\n",
    "    Output('map-id', \"children\"),\n",
    "    [Input('datatable-id', \"derived_viewport_data\"),\n",
    "     Input('datatable-id', 'derived_viewport_selected_rows')])\n",
    "def update_map(viewData, row_ids):\n",
    "    dff = pd.DataFrame.from_dict(viewData)\n",
    "    # Sets first num in row_ids array to the last selected row ID\n",
    "    row_ids[0] = row_ids[len(row_ids) - 1]\n",
    "    return [\n",
    "        # row_ids[0] gets last selected row from DataTable and parameters 13 and 14 are long and lat of rows\n",
    "        dl.Map(style={'width': '1000px', 'height': '500px'}, center=[dff.iloc[row_ids[0],13],dff.iloc[row_ids[0],14]], zoom=10, children=[\n",
    "            dl.TileLayer(id=\"base-layer-id\"),\n",
    "            # Marker with tool tip and popup\n",
    "            dl.Marker(position=[dff.iloc[row_ids[0],13],dff.iloc[row_ids[0],14]], children=[\n",
    "                # parameter num 4 gives the breed of animal\n",
    "                dl.Tooltip(dff.iloc[row_ids[0],4]),\n",
    "                dl.Popup([\n",
    "                    html.H1(\"Animal Name\"),\n",
    "                    # parameter 9 gives animal's name\n",
    "                    html.P(dff.iloc[row_ids[0],9])\n",
    "                ])\n",
    "            ])\n",
    "        ])\n",
    "    ]\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "031771a9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
